src/main/java/org/opensearch/knn/index/mapper/ModeBasedResolver.java

```diff
@@ -0,0 +1,158 @@
+import org.opensearch.knn.index.engine.MethodComponentContext;
+import org.opensearch.knn.index.engine.faiss.QFrameBitEncoder;
+import org.opensearch.knn.index.query.rescore.RescoreContext;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+
+import static org.opensearch.knn.common.KNNConstants.METHOD_ENCODER_PARAMETER;
+import static org.opensearch.knn.common.KNNConstants.METHOD_HNSW;
+import static org.opensearch.knn.common.KNNConstants.METHOD_PARAMETER_EF_CONSTRUCTION;
+import static org.opensearch.knn.common.KNNConstants.METHOD_PARAMETER_EF_SEARCH;
+import static org.opensearch.knn.common.KNNConstants.METHOD_PARAMETER_M;
+
+/**
+ * Class contains the logic to make parameter resolutions based on the {@link Mode} and {@link CompressionLevel}.
+ */
+public final class ModeBasedResolver {
+
+    public static final ModeBasedResolver INSTANCE = new ModeBasedResolver();
+
+    private static final CompressionLevel DEFAULT_COMPRESSION_FOR_MODE_ON_DISK = CompressionLevel.x32;
+    private static final CompressionLevel DEFAULT_COMPRESSION_FOR_MODE_IN_MEMORY = CompressionLevel.x1;
+    private final static Set<CompressionLevel> SUPPORTED_COMPRESSION_LEVELS = Set.of(
+        CompressionLevel.x1,
+        CompressionLevel.x2,
+        CompressionLevel.x8,
+        CompressionLevel.x16,
+        CompressionLevel.x32
+    );
+
+    private ModeBasedResolver() {}
+
+    /**
+     * Based on the provided {@link Mode} and {@link CompressionLevel}, resolve to a {@link KNNMethodContext}
+     *
+     * @param mode {@link Mode}
+     * @param compressionLevel {@link CompressionLevel}
+     * @return {@link KNNMethodContext}
+     */
+    public KNNMethodContext resolveKNNMethodContext(Mode mode, CompressionLevel compressionLevel) {
+        CompressionLevel resolvedCompressionLevel = resolveCompressionLevel(mode, compressionLevel);
+        MethodComponentContext encoderContext = resolveEncoder(resolvedCompressionLevel);
+
+        KNNEngine knnEngine = Mode.ON_DISK == mode || encoderContext != null ? KNNEngine.FAISS : KNNEngine.DEFAULT;
+
+        if (encoderContext != null) {
+            return new KNNMethodContext(
+                knnEngine,
+                SpaceType.DEFAULT,
+                new MethodComponentContext(
+                    METHOD_HNSW,
+                    Map.of(
+                        METHOD_PARAMETER_M,
+                        KNNSettings.INDEX_KNN_DEFAULT_ALGO_PARAM_M,
+                        METHOD_PARAMETER_EF_CONSTRUCTION,
+                        KNNSettings.INDEX_KNN_DEFAULT_ALGO_PARAM_EF_CONSTRUCTION,
+                        METHOD_PARAMETER_EF_SEARCH,
+                        KNNSettings.INDEX_KNN_DEFAULT_ALGO_PARAM_EF_SEARCH,
+                        METHOD_ENCODER_PARAMETER,
+                        encoderContext
+                    )
+                )
+            );
+        }
+
+        if (knnEngine == KNNEngine.FAISS) {
+            return new KNNMethodContext(
```

Comment:
The last two return statements look alike, then why do we need this condition ?

Reply:
Good catch - I updated. Difference is that nmslib does not take ef_search as index param while faiss does